src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -1054,6 +1120,258 @@ get_rebalance_progress(PG_FUNCTION_ARGS)
+	}
+
+	appendStringInfoString(query, "))");
+	appendStringInfoString(
+		query,
+		" SELECT shard_id, coalesce(pg_total_relation_size(tables.relid),0)"
+
+		/* for each shard in shardIds */
+		" FROM shard_names"
+
+		/* check if its name can be found in pg_class, if so return size */
+		" LEFT JOIN"
+		" (SELECT c.oid AS relid, c.relname, n.nspname"
+		" FROM pg_class c JOIN pg_namespace n ON n.oid = c.relnamespace) tables"
+		" ON tables.relname = shard_names.table_name AND"
+		" tables.nspname = shard_names.schema_name ");
+
+	PGresult *result = NULL;
+	int queryResult = ExecuteOptionalRemoteCommand(connection, query->data, &result);
+	if (queryResult != RESPONSE_OKAY)
+	{
+		ereport(ERROR, (errcode(ERRCODE_CONNECTION_FAILURE),
+						errmsg("cannot get the size because of a connection error")));
+	}
+
+	int rowCount = PQntuples(result);
+	int colCount = PQnfields(result);
+
+	/* This is not expected to ever happen, but we check just to be sure */
+	if (colCount < 2)
+	{
+		ereport(ERROR, (errmsg("unexpected number of columns returned by: %s",
+							   query->data)));
+	}
+
+	HASHCTL info = {
+		.keysize = sizeof(uint64),
+		.entrysize = sizeof(ShardStatistics),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardStatistics = hash_create("ShardStatisticsHash", 32, &info,
+										HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	for (int rowIndex = 0; rowIndex < rowCount; rowIndex++)
+	{
+		char *shardIdString = PQgetvalue(result, rowIndex, 0);
+		uint64 shardId = pg_strtouint64(shardIdString, NULL, 10);
+		char *sizeString = PQgetvalue(result, rowIndex, 1);
+		uint64 totalSize = pg_strtouint64(sizeString, NULL, 10);
+
+		ShardStatistics *statistics =
+			hash_search(shardStatistics, &shardId, HASH_ENTER, NULL);
+		statistics->totalSize = totalSize;
+	}
+
+	PQclear(result);
+
+	bool raiseErrors = true;
+	ClearResults(connection, raiseErrors);
+
+	return shardStatistics;
+}
+
+
+/*
+ * GetShardMoveShardIdsByWorker groups the shard ids in the provided steps by
+ * worker. It returns a hashmap that contains a set of these shard ids.
+ */
+static HTAB *
+GetShardMoveShardIdsByWorker(PlacementUpdateEventProgress *steps, int stepCount,
+							 bool fromSource)
+{
+	HASHCTL info = {
+		.keysize = sizeof(WorkerHashKey),
+		.entrysize = sizeof(WorkerShardIds),
+		.hcxt = CurrentMemoryContext
+	};
+
+	HTAB *shardsByWorker = hash_create("GetRebalanceStepsByWorker", 32, &info,
+									   HASH_ELEM | HASH_CONTEXT | HASH_BLOBS);
+
+	for (int stepIndex = 0; stepIndex < stepCount; stepIndex++)
+	{
+		PlacementUpdateEventProgress *step = &(steps[stepIndex]);
+
+		AddToWorkerShardIdSet(shardsByWorker, step->sourceName, step->sourcePort,
+							  step->shardId);
+
+		if (step->progress == REBALANCE_PROGRESS_WAITING)
+		{
+			/*
+			 * shard move has not started so we don't need target stats for
+			 * this shard
+			 */
+			continue;
+		}
+
+		AddToWorkerShardIdSet(shardsByWorker, step->targetName, step->targetPort,
+							  step->shardId);
```

Comment:
as I read this we skip the target shard if we know we haven't started the move. Wouldn't it make sense to also skip the source size if we know that the shard has already been completely moved? From the test I read (left a comment there as well) that if the shard has been moved we read a source size of 0 anyway, which makes sense to me.

If reading the source size for shards that are not there anymore (moved) is expected to result in a size of 0 I think this loop might be more readable with with two checks if we need to add them to source and target instead of having a break in the loop.

That said, it might be better nevertheless already to no continue to loop but check for the condition on which we want to add the shard to the set. eg:
```suggestion
		AddToWorkerShardIdSet(shardsByWorker, step->sourceName, step->sourcePort,
							  step->shardId);

		if (step->progress != REBALANCE_PROGRESS_WAITING)
		{
			/* shard either moved completely or is in progress, we want their current size on the target */
			AddToWorkerShardIdSet(shardsByWorker, step->targetName, step->targetPort,
								  step->shardId);
		}
```

It is generally easier to understand positive tests, when to add something, compared to a `continue` in a loop

Reply:
Like I said in the other comment, we don't necessarily read a source size of 0, because of `citus.defer_drop_after_shard_move`.